### ISSUE TITLE
Unable to have connectivity on GKE from control plane to pods with kube-proxy replacement enabled

### ISSUE THREAD

**Issue by hightoxicity** [cilium/cilium#12679]
## Support

Hi,

We have cilium 1.8.1 deployed into a private GKE cluster and we have some admissions webhooks + control plane that can not reach a pod IP, this is also the case from metrics server that is in-box with GKE.
The control plane is setup on 172.16.0.0/28 (managed by GKE), nodes on 10.1.0.0/20 and pods on 10.233.64.0/18, firewall rules (GCP fw rules) are setup from the control plane subnet to be able to reach all the previous targeted ports on host VPC subnetwork.

The workaround I use to have admission webhooks working consists in setting target webhooks pods on host net then control plane can reach the node IP + port, but I have no workaround for metrics server since its spec is managed by GKE.

Without talking about Cilium, reacheability is fixed only by a FW rule in GCP, look at:

https://github.com/kubernetes/kubernetes/issues/79739

They say that Google is using --enable-aggregator-routing=true to resolve endpoints IP directly so it is something expected that the control plane can talk to pods in GKE.

With cilium, It seems that something is preventing control plane (172.16.0.0/28) to talk to my pod subnet (10.233.64.0/18), is it something blocked by design into Cilium or shall I have to configure something to allow that?

Great thanks.

**Comment by hightoxicity** [comment]
Here is some details about my cilium helm deployment:

```
cilium:
  enabled: true
  config:
    ipam: kubernetes
  global:
    # prometheus enables
    prometheus:
      enabled: true
      port: 9092
      serviceMonitor:
        enabled: true
    # operatorPrometheus enables
    operatorPrometheus:
      enabled: true
      port: 6942
    hostServices:
      enabled: true
      protocols: tcp
    cni:
      binPath: /home/kubernetes/bin
    gke:
      enabled: true
    nativeRoutingCIDR: 10.233.64.0/18
    nodeinit:
      enabled: true
    psp:
      enabled: true
    hubble:
      enabled: true
      ui:
        enabled: true
      relay:
        enabled: true
      metrics:
        enabled:
        - dns:query;ignoreAAAA
        - drop
        - tcp
        - flow
        - port-distribution
        - icmp
        - http
        port: 9095
        serviceMonitor:
          enabled: true
  nodeinit:
    reconfigureKubelet: true
    removeCbrBridge: true

```

**Comment by aanm** [comment]
@hightoxicity what's the output of `kubectl get pods --all-namespaces -o wide`?

**Comment by hightoxicity** [comment]
```
kubectl get pods -n cilium -o wide
NAME                              READY   STATUS    RESTARTS   AGE    IP              NODE                                                NOMINATED NODE   READINESS GATES
cilium-cbh25                      1/1     Running   0          174m   10.1.0.2        gke-bel1-c1-custom-reg-pv-common-p1-9417c193-jc3n   <none>           <none>
cilium-chzjl                      1/1     Running   0          174m   10.1.0.4        gke-bel1-c1-custom-reg-pv-common-p1-b1fca78f-6kwd   <none>           <none>
cilium-jsh2x                      1/1     Running   0          174m   10.1.0.55       gke-bel1-c1-custom-reg-pv-pio-p1-46168f93-kncp      <none>           <none>
cilium-mqt8v                      1/1     Running   0          174m   10.1.15.192     gke-bel1-c1-custom-reg-pv-lp-job-p1-24c145af-w70g   <none>           <none>
cilium-nlc2g                      1/1     Running   0          174m   10.1.0.3        gke-bel1-c1-custom-reg-pv-common-p1-f4d2719f-sxjx   <none>           <none>
cilium-node-init-9qcxm            1/1     Running   0          174m   10.1.0.3        gke-bel1-c1-custom-reg-pv-common-p1-f4d2719f-sxjx   <none>           <none>
cilium-node-init-kqgc9            1/1     Running   0          174m   10.1.0.55       gke-bel1-c1-custom-reg-pv-pio-p1-46168f93-kncp      <none>           <none>
cilium-node-init-ktdtm            1/1     Running   0          174m   10.1.15.192     gke-bel1-c1-custom-reg-pv-lp-job-p1-24c145af-w70g   <none>           <none>
cilium-node-init-pf5hq            1/1     Running   0          174m   10.1.0.4        gke-bel1-c1-custom-reg-pv-common-p1-b1fca78f-6kwd   <none>           <none>
cilium-node-init-xs4wt            1/1     Running   0          174m   10.1.0.2        gke-bel1-c1-custom-reg-pv-common-p1-9417c193-jc3n   <none>           <none>
cilium-operator-cfc69df7f-jkvth   1/1     Running   0          176m   10.1.0.4        gke-bel1-c1-custom-reg-pv-common-p1-b1fca78f-6kwd   <none>           <none>
hubble-relay-6555c76d6-mmqb8      1/1     Running   0          4d1h   10.233.67.238   gke-bel1-c1-custom-reg-pv-pio-p1-46168f93-kncp      <none>           <none>
hubble-ui-5ddff94674-rsss5        1/1     Running   0          4d1h   10.233.65.18    gke-bel1-c1-custom-reg-pv-common-p1-f4d2719f-sxjx   <none>
```

**Comment by hightoxicity** [comment]
If I set:

> kubeProxyReplacement: disabled

It repairs connectivity between GKE control plane and POD ips

**Comment by brb** [comment]
@hightoxicity What was `cilium status` before you disabled the kube-proxy replacement?

**Comment by hightoxicity** [comment]
```
kubectl exec -it -n cilium cilium-29dtd -- /bin/sh
# cilium status
KVStore:                                                          Ok   Disabled
Kubernetes:                                                       Ok   1.16+ (v1.16.12-gke.3) [linux/amd64]
Kubernetes APIs:                                                  ["CustomResourceDefinition", "cilium/v2::CiliumClusterwideNetworkPolicy", "cilium/v2::CiliumEndpoint", "cilium/v2::CiliumNetworkPolicy", "cilium/v2::CiliumNode", "core/v1::Endpoint", "core/v1::Namespace", "core/v1::Node", "core/v1::Pods", "core/v1::Service", "networking.k8s.io/v1::NetworkPolicy"]
KubeProxyReplacement:                                             Probe   [eth0 (DR)]   [NodePort (SNAT, 30000-32767, XDP: DISABLED), HostPort, ExternalIPs, HostReachableServices (TCP, UDP), SessionAffinity]
Cilium:                                                           Ok      OK
NodeMonitor:                                                      Listening for events on 16 CPUs with 64x4096 of shared memory
Cilium health daemon:                                             Ok   
IPAM:                                                             IPv4: 34/255 allocated from 10.233.66.0/24, 
Masquerading:                                                     BPF   [eth0]   10.233.64.0/18
Controller Status:                                                177/177 healthy
Proxy Status:                                                     OK, ip 10.233.66.105, 0 redirects active on ports 10000-20000
Hubble:                                                           Ok              Current/Max Flows: 4096/4096 (100.00%), Flows/s: 122.93   Metrics: Ok
Cluster health:                                                   0/5 reachable   (2020-07-27T16:29:20Z)
  Name                                                            IP              Reachable   Endpoints reachable
  gke-bel1-c1-custom-reg-pv-common-p1-b1fca78f-6kwd (localhost)   10.1.0.4        false       false
  gke-bel1-c1-custom-reg-pv-common-p1-9417c193-jc3n               10.1.0.2        false       false
  gke-bel1-c1-custom-reg-pv-common-p1-f4d2719f-sxjx               10.1.0.3        false       false
  gke-bel1-c1-custom-reg-pv-lp-job-p1-24c145af-w70g               10.1.15.192     false       false
  gke-bel1-c1-custom-reg-pv-pio-p1-46168f93-kncp                  10.1.0.55       false       false
```

**Comment by hightoxicity** [comment]
Connectivity test result:

```
kubectl get pods -n cilium
NAME                                                     READY   STATUS    RESTARTS   AGE
cilium-29dtd                                             1/1     Running   0          9m21s
cilium-4m7b2                                             1/1     Running   0          9m20s
cilium-86pcq                                             1/1     Running   0          9m21s
cilium-crhfm                                             1/1     Running   0          9m17s
cilium-g5tqd                                             1/1     Running   0          9m16s
cilium-node-init-68v6l                                   1/1     Running   0          9m17s
cilium-node-init-6rf8d                                   1/1     Running   0          9m21s
cilium-node-init-9g2dl                                   1/1     Running   0          9m16s
cilium-node-init-fwjcn                                   1/1     Running   0          9m21s
cilium-node-init-z4qw2                                   1/1     Running   0          9m26s
cilium-operator-cfc69df7f-qc4tt                          1/1     Running   0          9m29s
echo-a-58dd59998d-xs27s                                  1/1     Running   0          2m59s
echo-b-865969889d-8m9pz                                  1/1     Running   0          2m58s
echo-b-host-659c674bb6-sm8pw                             1/1     Running   0          2m58s
host-to-b-multi-node-clusterip-6fb94d9df6-2vld6          1/1     Running   0          2m58s
host-to-b-multi-node-headless-7c4ff79cd-ch2ld            1/1     Running   0          2m58s
hubble-relay-6555c76d6-mmqb8                             1/1     Running   0          4d2h
hubble-ui-5ddff94674-rsss5                               1/1     Running   0          4d2h
pod-to-a-5c8dcf69f7-npgzl                                1/1     Running   0          2m57s
pod-to-a-allowed-cnp-75684d58cc-jp22z                    1/1     Running   0          2m58s
pod-to-a-external-1111-669ccfb85f-8cwvl                  1/1     Running   0          2m56s
pod-to-a-l3-denied-cnp-7b8bfcb66c-8pcdt                  1/1     Running   0          2m57s
pod-to-b-intra-node-74997967f8-bp9ll                     1/1     Running   0          2m57s
pod-to-b-intra-node-nodeport-775f967f47-2m6jb            1/1     Running   0          2m57s
pod-to-b-multi-node-clusterip-587678cbc4-x65qt           1/1     Running   0          2m57s
pod-to-b-multi-node-headless-574d9f5894-jqh2f            1/1     Running   0          2m57s
pod-to-b-multi-node-nodeport-7944d9f9fc-llg8g            1/1     Running   0          2m57s
pod-to-external-fqdn-allow-google-cnp-6dd57bc859-whnsh   1/1     Running   0          2m56s
```

**Comment by tgraf** [comment]
@hightoxicity Can you still reproduce this with the new native GKE support documented here?

https://cloud.google.com/blog/products/containers-kubernetes/bringing-ebpf-and-cilium-to-google-kubernetes-engine

**Comment by hightoxicity** [comment]
That is something engineered by Google, that is to say something you can turn on only creating a new cluster, not on existing cluster (like enabling rbac for groups), maybe they are thinking we are about to switch our clusters every 2 monthes, one day maybe I willl be able to migrate everything from one cluster to another but I can not just for now.
I am also still waiting for support in the tf gke module (https://github.com/terraform-google-modules/terraform-google-kubernetes-engine) but it will appear soon I think.

I think everything should be easier without this GKE blackbox, so having choice between testing the new dataplane v2 or leaving GKE and manage everything with something like kubespray (and a more opiniated deployment), I am seriously considering the second option.

If I decide to stay, I will come back to you to provide you my inputs, but I was scared to let load balancing done by iptable rules, that is the reason why Cilium was a great option.

THX

**Comment by fallard84** [comment]
We have the same issue with Cilium 1.9 running in GKE 1.17.13-gke.2600. Everything seems to work fine except for cert-manager admission webhook. We have a firewall rule that allows our master CIDR to access the webhook pod but with Cilium it doesn't work. Looking at Hubble everything seems fine (traffic from the master CIDR are `forwarded`) but we receive the following error when trying to create a resource that invoke the webhook: 
`Error from server (InternalError): error when creating "wildcard-tls.yaml": Internal error occurred: failed calling webhook "webhook.cert-manager.io": the server is currently unable to handle the request`

This is the same error message I would get without Cilium and with no firewall rule that allows the master to reach the pod.

**Comment by brb** [comment]
@fallard84 @hightoxicity Could you check with hubble and tcpdump what happens with each request? Does it reach a pod node?

**Comment by roko99** [comment]
Hi all,

Faced with same issue on GKE private cluster. Cilium version is 1.9.5 and k8s - 1.18.16-gke.502.
Hubble - packets are forwarded and tcpdump - packets reach pod node. But according to tcpdump on pod node, probably master doesn't receive ACKs. I can see SYN from master, SYN-ACK are sent from node, but nothing after this.
[hubble_UI.txt](https://github.com/cilium/cilium/files/6351288/hubble_UI.txt)
[hubble_obs_cli.txt](https://github.com/cilium/cilium/files/6351289/hubble_obs_cli.txt)
[cilium_dump.txt](https://github.com/cilium/cilium/files/6351290/cilium_dump.txt)

**Comment by brb** [comment]
@roko99 Thanks. Could you try to set `bpf.masquerade=false`?

**Comment by roko99** [comment]
@brb, `bpf.masquerade=false` resolves issue.

`cilium bpf ipcache get $CONTROL_PLANE_IP` returns `10.1.0.5 maps to identity 2 0 0.0.0.0`.

but is using iptables masquerading our target?

**Comment by brb** [comment]
What happens is that the reply gets unexpectedly SNAT-ed, as from the IPCache we think the the $CONTROL_PLANE_IP is outside the cluster (aka WORLD_ID). So you are hitting https://github.com/cilium/cilium/issues/12544.

Just to confirm, the control plane does not run Cilium, right?

**Comment by roko99** [comment]
Yes, it's managed by GKE.
But as far as I can control it`s IP range - `ipMasqAgent` helps me. So no issue here for my case, sorry.

**Comment by errordeveloper** [comment]
@brb can we close this as duplicate of #12544?

**Comment by brb** [comment]
@errordeveloper Yep. Closing in favor of #12544.
